pallets/pools/src/lib.rs

```diff
@@ -42,22 +42,54 @@ use sp_std::vec::Vec;
 pub struct PoolDetails<AccountId, CurrencyId, EpochId, Balance, Rate, MetaSize, Weight>
 where
 	MetaSize: Get<u32> + Copy,
+	Rate: FixedPointNumber<Inner = Balance>,
+	Balance: FixedPointOperand,
 {
 	pub owner: AccountId,
 	pub currency: CurrencyId,
-	pub tranches: Tranches<Balance, Rate, Weight>, // ordered junior => senior
+	pub tranches: Tranches<Balance, Rate, Weight, CurrencyId>, // ordered junior => senior
 	pub current_epoch: EpochId,
 	pub last_epoch_closed: Moment,
 	pub last_epoch_executed: EpochId,
-	pub max_reserve: Balance,
-	pub available_reserve: Balance,
-	pub total_reserve: Balance,
+	pub reserve: ReserveDetails<Balance>,
 	pub metadata: Option<BoundedVec<u8, MetaSize>>,
 	pub min_epoch_time: Moment,
 	pub challenge_time: Moment,
 	pub max_nav_age: Moment,
 }
 
+#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct ReserveDetails<Balance> {
+	pub max_reserve: Balance,
+	pub available_reserve: Balance,
+	pub total_reserve: Balance,
+}
+
+impl<AccountId, CurrencyId, EpochId, Balance, Rate, MetaSize, Weight>
+	PoolDetails<AccountId, CurrencyId, EpochId, Balance, Rate, MetaSize, Weight>
+where
+	MetaSize: Get<u32> + Copy,
+	Rate: FixedPointNumber<Inner = Balance>,
+	Balance: FixedPointOperand,
+	EpochId: BaseArithmetic,
+{
+	pub fn end_epoch(&mut self, now: Moment) -> DispatchResult {
+		self.current_epoch += One::one();
+		self.last_epoch_closed = now;
+		// TODO: Remove and set state rather to EpochClosing or similar
+		// Set available reserve to 0 to disable originations while the epoch is closed but not executed
+		self.reserve.available_reserve = Zero::zero();
+
+		Ok(())
+	}
+
+	fn start_epoch(&mut self) -> DispatchResult {
```

Comment:
I think this naming is confusing. The next epoch starts immediately when `current_epoch` is incremented. New orders can already come in for this new epoch while the previous epoch is still in a submission period, as well as repayments (just not originations as those are blocked by the `available_reserve` being 0). I'm not sure what the best name for this is though, the best I came come up with is something like `process_executed_epoch`.